1.Save the Saturday, June 17th, 2017 data to a csv file. You can find it here: http://web.mta.info/developers/turnstile.html and read it in using read_csv. View the head of the dataframe.

In [14]:
#insert 1
import pandas as pd  

data = pd.read_csv('june17.csv')
data.head()

C/A  UNIT       SCP STATION LINENAME DIVISION        DATE      TIME  \
0  A002  R051  02-00-00   59 ST  NQR456W      BMT  06/10/2017  00:00:00   
1  A002  R051  02-00-00   59 ST  NQR456W      BMT  06/10/2017  04:00:00   
2  A002  R051  02-00-00   59 ST  NQR456W      BMT  06/10/2017  08:00:00   
3  A002  R051  02-00-00   59 ST  NQR456W      BMT  06/10/2017  12:00:00   
4  A002  R051  02-00-00   59 ST  NQR456W      BMT  06/10/2017  16:00:00   

      DESC  ENTRIES  \
0  REGULAR  6215258   
1  REGULAR  6215284   
2  REGULAR  6215318   
3  REGULAR  6215475   
4  REGULAR  6215841   

   EXITS                                                                 
0                                            2104297                     
1                                            2104303                     
2                                            2104337                     
3                                            2104417                     
4                                            2104465

2.View the dataframe columns. Is there extra whitespace at the end of any of the column names? Strip all of the whitespace.

In [15]:
#insert 2
data.columns = data.columns.str.strip()
data.columns  

Index(['C/A', 'UNIT', 'SCP', 'STATION', 'LINENAME', 'DIVISION', 'DATE', 'TIME',
       'DESC', 'ENTRIES', 'EXITS'],
      dtype='object')

3.In one line, determine how many rows are in your dataframe and if there are are any null entries.

In [16]:
#insert 3
data.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 197209 entries, 0 to 197208
Data columns (total 11 columns):
C/A         197209 non-null object
UNIT        197209 non-null object
SCP         197209 non-null object
STATION     197209 non-null object
LINENAME    197209 non-null object
DIVISION    197209 non-null object
DATE        197209 non-null object
TIME        197209 non-null object
DESC        197209 non-null object
ENTRIES     197209 non-null int64
EXITS       197209 non-null int64
dtypes: int64(2), object(9)
memory usage: 16.6+ MB


4.In one line, print out the unique STATION names. How many are there? (Hint: use set).

In [17]:
#insert 4
data.STATION.unique()

array(['59 ST', '5 AV/59 ST', '57 ST-7 AV', '49 ST', 'TIMES SQ-42 ST',
       '34 ST-HERALD SQ', '28 ST', '23 ST', '14 ST-UNION SQ', '8 ST-NYU',
       'PRINCE ST', 'CANAL ST', 'CITY HALL', 'CORTLANDT ST', 'RECTOR ST',
       'WHITEHALL S-FRY', 'DELANCEY/ESSEX', 'BOWERY', 'CHAMBERS ST',
       'FULTON ST', 'BROAD ST', '7 AV', 'PARK PLACE', 'BOTANIC GARDEN',
       'PROSPECT PARK', 'PARKSIDE AV', 'CHURCH AV', 'BEVERLEY ROAD',
       'CORTELYOU RD', 'NEWKIRK PLAZA', 'AVENUE H', 'AVENUE J',
       'AVENUE M', 'KINGS HWY', 'AVENUE U', 'NECK RD', 'SHEEPSHEAD BAY',
       'BRIGHTON BEACH', 'OCEAN PKWY', 'BOROUGH HALL', 'JAY ST-METROTEC',
       'DEKALB AV', 'ATL AV-BARCLAY', 'UNION ST', '4AV-9 ST', '25 ST',
       '36 ST', '45 ST', '77 ST', '86 ST', 'BAY RIDGE-95 ST', '8 AV',
       'FT HAMILTON PKY', 'NEW UTRECHT AV', '18 AV', '20 AV', 'BAY PKWY',
       '9 AV', '50 ST', '55 ST', '71 ST', '79 ST', '25 AV', 'BAY 50 ST',
       'CONEY IS-STILLW', 'W 8 ST-AQUARIUM', '6 AV', '3 AV', '1 AV',
   

5.Okay, so we understand what station represents. But what the heck are C/A, UNIT, and SCP? Keep in mind that in the larger stations, you might have multiple areas within one station that look like this:

<img src="image.jpg" style="width: 300px;"/>

Further complicating things, there are a few station names like 14TH ST that refer to more than one station location along that street.

This data set is not very well documented. Welcome to the joys of real world data science!!!

Read the following two links carefully to see other people's confusion and what information they have been able to gather:

https://groups.google.com/forum/#!topic/mtadeveloperresources/AMVx2WUY9iI

https://groups.google.com/forum/#!searchin/mtadeveloperresources/%22remote$20unit%22%7Csort:relevance/mtadeveloperresources/z8l3ZU9cY6Y/OFlHGkFAimQJ

What did Patrick O'Hara and Katie Toth have to say?

In [18]:
#insert 5
print('Some stations in MTA datafile have very strange name that do not make sense in the real life. So it is hard to refer and find a specific station.')
print('Also it is hard to understand what is a control area and other terms that are used in MTA datafile. ')

Some stations in MTA datafile have very strange name that do not make sense in the real life. So it is hard to refer and find a specific station.
Also it is hard to understand what is a control area and other terms that are used in MTA datafile. 


6.To simplify our calculations right now, let's only look at ridership on Monday, June 12th, 2017. In one line, create a dataset called monday that just includes the dates listed as 06/12/2017.

Then, use the command "monday = monday.reset_index(drop=True)" to reset the index.

In [19]:
#insert 6
monday=data[data.DATE=='06/12/2017']
monday=monday.reset_index(drop=True) 
monday.head()

C/A  UNIT       SCP STATION LINENAME DIVISION        DATE      TIME  \
0  A002  R051  02-00-00   59 ST  NQR456W      BMT  06/12/2017  00:00:00   
1  A002  R051  02-00-00   59 ST  NQR456W      BMT  06/12/2017  04:00:00   
2  A002  R051  02-00-00   59 ST  NQR456W      BMT  06/12/2017  08:00:00   
3  A002  R051  02-00-00   59 ST  NQR456W      BMT  06/12/2017  12:00:00   
4  A002  R051  02-00-00   59 ST  NQR456W      BMT  06/12/2017  16:00:00   

      DESC  ENTRIES    EXITS  
0  REGULAR  6217322  2104727  
1  REGULAR  6217332  2104729  
2  REGULAR  6217372  2104813  
3  REGULAR  6217540  2105024  
4  REGULAR  6217825  2105081

7.Create a dictionary called monday_dict mapping the tuple (C/A, UNIT,SCP, STATION) to a list of tuples containing (TIME,EXITS).

For example, the key ('A002', 'R051', '02-00-00','59 ST') in your dictionary should give the value:

[('00:00:00', 2104727),
 ('04:00:00', 2104729),
 ('08:00:00', 2104813),
 ('12:00:00', 2105024),
 ('16:00:00', 2105081),
 ('20:00:00', 2105134)]

In [20]:
#insert 7

monday_dict={}
for i in range(len(monday.index)):
    tuple1 = ((monday['C/A'][i]), monday.UNIT[i], monday.SCP[i], monday.STATION[i])
    if tuple1 not in monday_dict.keys():
        monday_dict[tuple1]=[(monday.TIME[i],monday.EXITS[i])]
    else:
        monday_dict[tuple1].append((monday.TIME[i],monday.EXITS[i]))
        
monday_dict[('A002', 'R051', '02-00-00','59 ST')]


[('00:00:00', 2104727),
 ('04:00:00', 2104729),
 ('08:00:00', 2104813),
 ('12:00:00', 2105024),
 ('16:00:00', 2105081),
 ('20:00:00', 2105134)]

8.Note: as you get more comfortable with pandas, you will rely more on built-in pandas capabilities and less on creating dictionaries from scratch. For example, run the following lines of code to verify that this dataframe gives you the same info that you got in Exercise 7.

In [21]:
x=monday.groupby(['C/A', 'UNIT', 'SCP', 'STATION']).groups #the keys are the station pairs and the values are the row indices
flds = ['TIME','EXITS'] #we want to make values out of the remaining columns
monday_exits = {key: monday.loc[val,flds] for key, val in x.items()}
monday_exits[('A002', 'R051', '02-00-00','59 ST')]

TIME    EXITS
0  00:00:00  2104727
1  04:00:00  2104729
2  08:00:00  2104813
3  12:00:00  2105024
4  16:00:00  2105081
5  20:00:00  2105134

9.What the heck do these exit numbers represent? We can assume that they represent some sort of counter on the turnstile. We can subtract the exits at t=20 minus t=0 to figure out how many riders exited this turnstile between midnight and 8 pm.

Can we calculate 20 - 0 for all of the stations? No! Unfortunately, the time increments aren't standardized. Many of the times are listed in 4 hour intervals of midnight, 4 am, 8 am, etc but not all. Make a frequency chart of the times for the monday dataframe from Exercise 6.



In [22]:
#insert 9

monday.TIME.value_counts()

00:00:00    2409
12:00:00    2408
20:00:00    2408
08:00:00    2407
04:00:00    2405
16:00:00    2405
17:00:00    1741
13:00:00    1740
01:00:00    1740
05:00:00    1740
09:00:00    1739
21:00:00    1738
06:00:00     110
22:00:00     110
02:00:00     110
10:00:00     110
14:00:00     110
18:00:00     110
03:00:00      39
15:00:00      39
23:00:00      39
19:00:00      39
07:00:00      39
11:00:00      39
00:22:00      37
08:22:00      37
16:22:00      37
12:22:00      37
04:22:00      37
20:22:00      37
            ... 
16:30:24       1
13:58:11       1
05:04:24       1
17:54:55       1
06:18:49       1
19:03:05       1
21:01:18       1
06:13:05       1
19:40:11       1
09:20:48       1
13:34:01       1
22:59:51       1
13:07:27       1
16:03:13       1
07:08:37       1
17:39:05       1
23:15:52       1
12:04:49       1
01:18:31       1
16:30:52       1
22:01:14       1
07:12:42       1
09:00:14       1
12:55:02       1
11:37:22       1
12:11:52       1
09:42:58       1
20:33:26      

10.Use your monday_dict from Exercise 7 to create a dictionary called monday_dict_diff that maps (C/A, UNIT, SCP, STATION) to the differences in exits between two consective times in order to see how many people exited through a turnstile within a given time interval.

For example, the key ('A002', 'R051', '02-00-00','59 ST') should map to:

[('04:00:00', 2),
 ('08:00:00', 84),
 ('12:00:00', 211),
 ('16:00:00', 57),
 ('20:00:00', 53)]

In [23]:
#insert 10
monday_dict_diff={} 

for key1,value1 in monday_dict.items(): 
     for i in range(0,len(monday_dict[key1])-1):
        if key1 not in monday_dict_diff.keys():
            monday_dict_diff[key1]=[(value1[i+1][0],value1[i+1][1]-value1[i][1])] 
        else:
            monday_dict_diff[key1].append((value1[i+1][0],value1[i+1][1]-value1[i][1])) 
                                          
f
print(monday_dict_diff[('A002', 'R051', '02-00-00','59 ST')])

NameError: name 'f' is not defined

11.Once again, note that pandas can do the above calculation for you. If you used the monday_exits dataframe from Exercise 8, run the following code to verify that you obtain what you just got in Exercise 10. It may take a bit longer to run.

In [ ]:
monday_diff = {}

for turnstile, slice_df in monday_exits.items():
    diff = slice_df['EXITS'].diff().values #calculates the exit difference between each entry
    monday_diff[turnstile] = pd.Series(diff, index=slice_df['TIME'].values) #maps turnstile to (time, exit difference) pairs
    for i,(key, val) in enumerate(monday_diff[turnstile].items()):
        try: #need a try because you'll get an error for NAN difference values since there is one less difference than the number of times
            monday_exits[turnstile].at[i,'EXITS'] = val
        except:
            continue
    monday_exits[turnstile] = monday_exits[turnstile].drop(0)    #drop the first time because you only calculate differences after the second time.                   
    
monday_exits[('A002', 'R051', '02-00-00', '59 ST')]

12.Find how many people in total passed through a given turnstile between the hours given for 06/12/2017. For example, if a turnstile is on increments of 0,4,8,...,20, then you want to find out how many total people passed through between midnight and 8 pm. Create a dictionary called total_turnstile that maps (C/A,UNIT,SCP,STATION) to total Monday ridership. Use your dictionary from Exercise 10.

Hint: the key ('A002', 'R051', '02-00-00', '59 ST') should map to 407 total riders.

In [ ]:
#insert 12
total_turnstile={}
for key,value in monday_dict_diff.items():
    total=0 
    for i in range(len(value)):
        total=total+value[i][1]
        total_turnstile[key]=total
print(total_turnstile[('A002', 'R051', '02-00-00', '59 ST')])


13.Use your work in Exercise 12 to create a dictionary called total_riders that maps the (C/A, UNIT, STATION) location to all riders that passed through ALL of the turnstiles in that area on Monday.

Hint 1: use the get command

Hint 2: ('A002', 'R051', '59 ST') should have 6704 total riders.

In [ ]:
#insert 13
total_riders={}

for key,value in total_turnstile.items():
    tuple1=key[0],key[1],key[3]
    if (tuple1) not in total_riders.keys():
        total_riders[tuple1]=[total_turnstile.get(key,0)]
    else:
        total_riders[tuple1].append(total_turnstile.get(key,0))
    
for key,value in total_riders.items():
    total_riders[key]=sum(value)
    
print(total_riders[('A002', 'R051', '59 ST')])   

#total_riders={}
#for key,value in total_turnstile.items():
#    tuple1=key[0],key[1],key[3] 
#    total_riders[tuple1]=total_riders.get(tuple1,0)+value
#print(total_riders[('A002', 'R051', '59 ST')]) 

    

14.Create a list of tuples called tuple_list using your total_riders dictionary so that you can sort the list in decending order to see busiest stations. Hint: the busiest area should be the PATH World Trade Center station.

In [ ]:
#insert 14
tuple_list=[]
for key,value in total_riders.items():
    tuple_list.append((value,key[2]))

print(sorted(tuple_list)[::-1][0])


15.Look at the three smallest riderships in your above tuple list. Do they make sense? What about your data caused negative numbers? Welcome to the world of buggy real world data! Remove these three items from your tuple list.

In [ ]:
#insert 15
print('Number of people coming in is larger than the number of people exiting.')
tuple_list=sorted(tuple_list)[::-1]

tuple_list.pop(-1)
tuple_list.pop(-1)
tuple_list.pop(-1)
print(tuple_list)

16.Create a barchart for the number of riders. You don't need to keep track of the names of each station, as that would get really messy to plot. Instead, you can just use a list comprehension to save the rider numbers to the y-values and give the stations the numbers 0,1,2,3,... for the x-values. You can use the following piece of code. Remember to add in axis labels and a title.

x = range(len(tuple_list))

y = [tuple_list[i][0] for i in range(len(tuple_list))]

plt.bar(x,y)

In [ ]:
#insert 16

import matplotlib.pyplot as plt 

x = range(len(tuple_list))
y = [tuple_list[i][0] for i in range(len(tuple_list))]
plt.bar(x,y) 



17.Perhaps we don't care about total ridership in a 24 period. Perhaps we only care about the morning commute. Let's assume that the majority of people arrive at their work location subway exit between 8 and 9 am.  

Go back up to your frequency chart in Exercise 9. Note that the most common morning commutes would be gained from subtracting 9 am - 5 am and 10 am - 6 am and 11 am - 7 am and 12pm - 8am, for the stations that are broken down by those specific four hour intervals.

As a dirty way to consider only the stations that are on these time intervals, save just the rows in your original monday dataframe that contain the TIME strings '05:00:00', '06:00:00',..., '12:00:00'. Remember you can do this by using OR statements using the "|" symbol and remembering parenthesis. Call this new dataframe "morning". Be sure to reset its index.

In [ ]:
#insert 17
monday=monday[(monday.TIME=='05:00:00')|(monday.TIME=='12:00:00')|(monday.TIME=='06:00:00')|
               (monday.TIME=='07:00:00')|(monday.TIME=='08:00:00')|(monday.TIME=='09:00:00')|
               (monday.TIME=='10:00:00')|(monday.TIME=='11:00:00')|(monday.TIME=='12:00:00')]
monday=monday.reset_index(drop=True) 
monday.head()


18.Now create a morning_dict by repeating your work in Exercise 7 on this morning dataframe. The dictionary should map the tuple (C/A, UNIT, SCP,STATION) to an inner dictionary containing {TIME: EXITS} for all Monday rides.

Hint: ('A002', 'R051', '02-00-00','59 ST') should map to {'08:00:00': 2104813, '12:00:00': 2105024}.

In [ ]:
#insert 18
monday_dict={}
for i in range(len(monday.index)):
        tuple1 = ((monday['C/A'][i], monday.UNIT[i], monday.SCP[i], monday.STATION[i]))
        monday_dict[tuple1]=monday_dict.get(tuple1,{})
        monday_dict[tuple1].update({monday.TIME[i]:monday.EXITS[i]})

print(monday_dict[('A002', 'R051', '02-00-00','59 ST')])


19.We're almost ready to subtract the two consecutive times for each station to get the morning commutes. That is, if there are exactly two times for each station. Find which stations keys in your dictionary above do not contain exactly two times and delete them from the morning dictionary.

How many keys are left in your morning_dict? There should be 4293.

In [ ]:
#insert 19
to_delete = []
for key,value in monday_dict.items():
    if len(value)!=2:
        to_delete.append(key)
for i in to_delete:
    del monday_dict[i]
    
print(len(monday_dict))


20.Create a dictionary called morning_differences where (C/A, UNIT, STATION) maps to an inner dictionary containing {SCP: Morning Exit Differences}.

For example, ('A002', 'R051', '59 ST') should map to:

{'02-00-00': 211,
 '02-00-01': 128,
 '02-03-00': 654,
 '02-03-01': 630,
 '02-03-02': 533,
 '02-03-03': 401,
 '02-03-04': 186,
 '02-03-05': 49,
 '02-03-06': 8,
 '02-05-00': 0,
 '02-05-01': 0,
 '02-06-00': 4}

In [ ]:
#insert 20
morning_differences={}
for key,value in monday_dict.items():
    key1=(key[0],key[1],key[3])
    scp=key[2]
    list1=[]
    for key2,value2 in value.items():
        list1.append(value2)
    diff=list1[1]-list1[0]
        
    if key1 not in morning_differences.keys():
        morning_differences[key1]={scp:diff}
    else:
        morning_differences[key1][scp]=diff 
        
morning_differences[('A002', 'R051', '59 ST')]

21.Create a dictionary called total_morning_diff that maps each (C/A, Unit, Station) location to the total morning riders passing through ALL of the turnstiles in that area.

Hint 1: see your work in Exercise 13.

Hint 2: ('A002', 'R051','59 ST') should have 2804 morning commuter exits.


In [ ]:
#insert 21
total_morning_diff={}
for key,value in morning_differences.items():
    total=[]
    for key1,value1 in value.items():
        total.append(value1)
    sum1=sum(total)
    total_morning_diff[key]=total_morning_diff.get(key,0)+sum1
    
total_morning_diff[('A002', 'R051','59 ST')]

22.Create a list of sorted tuples to see the most popular morning exits.

Hint:  ('R238', 'R046', 'GRD CNTRL-42 ST') should be the highest with 17,300 morning exits.

In [38]:
#insert 22
tuple1=[]
for key,value in total_morning_diff.items():
    tuple1.append((value,key))
print(sorted(tuple1)[::-1][0])


NameError: name 'total_morning_diff' is not defined

23.We notice that the ('R204', 'R043', 'WALL ST') stop is pretty high on the Monday morning list. How can we tell if this exit is more of a commuter or tourist exit? Maybe we'll see if the total number of people is higher on a weekday than a weekend. Go back to the original df dataframe. Save only the items that contain ('R204', 'R043', 'WALL ST') to a dataframe called wallstreet and create a list of tuples where the first entry is the date and the second entry is the total ridership. 

Hint: there should be 9936 riders on 06/15/2017.

In [43]:
#insert 23

sed=data.rename(columns={"C/A":"CA"})
sed=sed[(sed.CA=='R204')&(sed.UNIT=='R043')&(sed.STATION=='WALL ST')]
sed=sed.reset_index(drop=True) 
wallstreet=sed.rename(columns={"CA":"C/A"})
wallstreet
     
list1=[]
dic={}
for i in range(len(wallstreet.index)):
    key=wallstreet.DATE[i]
    if key not in dic.keys():
        dic[key]={wallstreet.SCP[i]:{wallstreet.TIME[i]:wallstreet.EXITS[i]}}
    if wallstreet.SCP[i] not in dic[key].keys():
        dic[key][wallstreet.SCP[i]]={wallstreet.TIME[i]:wallstreet.EXITS[i]}
    else:
        dic[key][wallstreet.SCP[i]][wallstreet.TIME[i]]=wallstreet.EXITS[i]

dic['06/12/2017']

          
wall_dict={}

for key,value in dic.items():
    for key1,value1 in value.items():
        list1=[]
        list2=[]
        for key2,value2 in value1.items():
            total=0
            list1.append(value1[key2])
        sums=[]
        for num in list1:
            for i in range(0,len(list1)-1):
                list2.append(list1[i+1]-list1[i])
            sums.append(sum(list2))
        wall_dict[key]=wall_dict.get(key,{})
        wall_dict[key].update({key1:sums[0]})
        
wall_dict['06/12/2017']
tuple1=[]
for key,value in wall_dict.items():
    list3=[]
    for key1,value1 in value.items():
        list3.append(value1)
    tuple1.append((key,sum(list3)))
 
print(tuple1)

                                            

[('06/10/2017', 11), ('06/11/2017', 1745), ('06/12/2017', 9397), ('06/13/2017', 9730), ('06/14/2017', 9501), ('06/15/2017', 9936), ('06/16/2017', 8778)]


24.Plot the list of tuples you made where date is the x-axis and ridership is the y-axis for this wall street station.

Hint: you'll first need to use datetime to convert the dates to datetime format.

In [ ]:
#insert 24
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
from datetime import datetime

date1=[]
rider=[]
for i in range(len(tuple1)):
    date=datetime.strptime(tuple1[i][0],"%m/%d/%Y")
    date1.append(date)
    rider.append(tuple1[i][1])



dates = np.array(date1)
riders = np.array(rider)
plt.plot(dates, riders,'.')
plt.show()

25.Now use datetime to convert the dates to "Mon", "Tues", etc and print the days of the week with corresponding ridership. Is this a commuter stop or a tourist stop?

In [ ]:
#insert 25
import datetime
y=[]
for i in range(len(tuple1)):
    y.append(datetime.datetime.strptime(tuple1[i][0],"%m/%d/%Y").strftime('%a'))
    y.append(tuple1[i][1])
print(y)
print('This is a commuter stop.')

In [4]:
import pandas as pd  

def Stations(MTA):
    MTA.columns = MTA.columns.str.strip()
    monday = MTA[(MTA.STATION == '59 ST') | (MTA.STATION =='57 ST') | (MTA.STATION == '5 AVE') | (MTA.STATION == '21 ST-QNSBRIDGE') | (MTA.STATION == '23 ST') | (MTA.STATION =='42 ST-PORT AUTH') | (MTA.STATION == 'CYPRESS AV') | (MTA.STATION == 'YORK ST') | (MTA.STATION == 'BOROUGH HALL') | (MTA.STATION == 'HOYT ST') | (MTA.STATION == 'NEVINS ST')]
    monday = monday.reset_index(drop=True)
    return(monday)



def time(MTA):
    evening=MTA[(MTA.TIME=='12:00:00')|(MTA.TIME=='13:00:00')|(MTA.TIME=='14:00:00')|
               (MTA.TIME=='15:00:00')|(MTA.TIME=='16:00:00')|(MTA.TIME=='17:00:00')|
               (MTA.TIME=='18:00:00')|(MTA.TIME=='19:00:00')|(MTA.TIME=='20:00:00')]
    evening=evening.reset_index(drop=True) 
    return(evening)

stations = Stations(pd.read_csv('june17.csv'))
times = time(stations)

def date(MTA):
    MTA=MTA[MTA.DATE=='06/16/2017']
    MTA=MTA.reset_index(drop=True)
    return(MTA)

droped=date(times)

def dic(MTA):
    MTA_dict={}
    for i in range(len(MTA.index)):
            tuple1 = ((MTA['C/A'][i], MTA.UNIT[i], MTA.SCP[i], MTA.STATION[i]))
            MTA_dict[tuple1]=MTA_dict.get(tuple1,{})
            MTA_dict[tuple1].update({MTA.TIME[i]:MTA.EXITS[i]})
    return(MTA_dict)
dic1=dic(droped)



def delete(MTA):
    to_delete = []
    for key,value in MTA.items():
        if len(value)!=2:
            to_delete.append(key)
    for i in to_delete:
        del MTA[i]
    return(MTA)

deleted_dict=delete(dic1)

def morningdif(mordict):
    morning_differences = {}
    for key,value in mordict.items():
        newkey = (key[0],key[1],key[3])
        val = []
        SCP = key[2]
        for key1, val1 in value.items():
            val.append(val1)
        dif = val[1]-val[0]
        if newkey not in morning_differences:
            morning_differences[newkey] = {SCP : dif}
        else:
            morning_differences[newkey][SCP] = dif
    return(morning_differences)

morningdiff=morningdif(deleted_dict)

def totalnum (morningdif):
    total_morning_diff={}
    for key,value in morningdif.items():
        total=[]
        for key1,value1 in value.items():
            total.append(value1)
            sum1=sum(total)
        total_morning_diff[key]=total_morning_diff.get(key,0)+sum1
    return (total_morning_diff)

total=totalnum(morningdiff)


def rank(totalnum):
    list1=[]
    for key2,value2 in totalnum.items():
        if key2 not in list1:
            list1.append((value2,key2))
    list1.sort(reverse=True)
    return(list1)
ranked=rank(total)

print(ranked)

[(4140, ('N062A', 'R010', '42 ST-PORT AUTH')), (3473, ('N060', 'R010', '42 ST-PORT AUTH')), (2190, ('N600', 'R302', '57 ST')), (1988, ('R132', 'R190', '23 ST')), (1890, ('R226', 'R131', '23 ST')), (1027, ('R131', 'R190', '23 ST')), (953, ('N530', 'R301', 'YORK ST')), (776, ('R227A', 'R131', '23 ST')), (692, ('R226A', 'R131', '23 ST')), (292, ('R404', 'R447', 'CYPRESS AV')), (287, ('R405', 'R447', 'CYPRESS AV')), (122, ('R227', 'R131', '23 ST'))]


In [25]:

import pandas as pd  

def Stations(MTA):
    MTA.columns = MTA.columns.str.strip()
    monday = MTA[(MTA.STATION == '59 ST') | (MTA.STATION =='57 ST') | (MTA.STATION == '5 AVE') | (MTA.STATION == '21 ST-QNSBRIDGE') | (MTA.STATION == '23 ST') | (MTA.STATION =='42 ST-PORT AUTH') | (MTA.STATION == 'CYPRESS AV') | (MTA.STATION == 'YORK ST') | (MTA.STATION == 'BOROUGH HALL') | (MTA.STATION == 'HOYT ST') | (MTA.STATION == 'NEVINS ST')]
    monday = monday.reset_index(drop=True)
    return(monday)



def time(MTA):
    morning=MTA[(MTA.TIME=='05:00:00')|(MTA.TIME=='12:00:00')|(MTA.TIME=='06:00:00')|
               (MTA.TIME=='07:00:00')|(MTA.TIME=='08:00:00')|(MTA.TIME=='09:00:00')|
               (MTA.TIME=='10:00:00')|(MTA.TIME=='11:00:00')|(MTA.TIME=='12:00:00')]
    morning=morning.reset_index(drop=True) 
    return(morning)

stations = Stations(pd.read_csv('june17.csv'))
times = time(stations)

def date(MTA):
    MTA=MTA[MTA.DATE=='06/10/2017']
    MTA=MTA.reset_index(drop=True)
    return(MTA)

droped=date(times)

def dic(MTA):
    MTA_dict={}
    for i in range(len(MTA.index)):
            tuple1 = ((MTA['C/A'][i], MTA.UNIT[i], MTA.STATION[i]))
            MTA_dict[tuple1]=MTA_dict.get(tuple1,{})
            MTA_dict[tuple1].update({MTA.TIME[i]:MTA.EXITS[i]})
    return(MTA_dict)
dic1=dic(droped)

def delete(MTA):
    to_delete = []
    for key,value in MTA.items():
        if len(value)!=2:
            to_delete.append(key)
    for i in to_delete:
        del MTA[i]
    return(MTA)

deleted_dict=delete(dic1)
 
print(deleted_dict[('R601A', 'R108', 'BOROUGH HALL')])


{'08:00:00': 1830, '12:00:00': 1830}
